### PR TITLE
add fillna and fillnull expressions and dataframe methods

### PR DESCRIPTION
**Pull request by FelixKleineBoesing**
Implements Issue #571

### CODE REVIEW

**Comment by FelixKleineBoesing** [comment]
@jaychia This can be reviewed and merged.

File: daft/dataframe/dataframe.py [hunk 1]
```diff
@@ -1120,6 +1120,36 @@ def __len__(self):
         )
         raise RuntimeError(message)
 
+    @DataframePublicAPI
```
Comment (jaychia) [inline daft/dataframe/dataframe.py:1123]: Let's not implement `fillnan` and `fillnull` on a DataFrame level?

You're correct that this `Only works if the columns have the same dtype.` - because of this, these functions will likely not work on most Dataframes.

This is almost never the case when working with Daft Dataframes because unlike a framework such as Pandas/Numpy, we deal with much more than just numerical data!

File: daft/expressions.py [hunk 2]
```diff
@@ -570,6 +570,29 @@ def cast(self, to: type) -> Expression:
...
+        return self.is_nan().if_else(val, self)
+
+    def fillnull(self, val: Any):
+        """Fills any NULL with the value of the val argument.
+
+        Args:
+            val: value which replaces NULL values
+
+        Returns:
+
```
Comment (jaychia) [inline daft/expressions.py:592]:
```suggestion
            Expression: new expression with all NULL values replaced with `val`
```

File: daft/expressions.py [hunk 3]
```diff
@@ -570,6 +570,29 @@ def cast(self, to: type) -> Expression:
...
+        Args:
+            val: value which replaces NaN values
+
+        Returns:
+
+        """
+        return self.is_nan().if_else(val, self)
+
+    def fillnull(self, val: Any):
```
Comment (jaychia) [inline daft/expressions.py:585]:
```suggestion
    def fillnull(self, val: Any) -> Expression:
```

File: daft/expressions.py [hunk 4]
```diff
@@ -570,6 +570,29 @@ def cast(self, to: type) -> Expression:
...
+        """Fills any NULL with the value of the val argument.
+
+        Args:
+            val: value which replaces NULL values
+
+        Returns:
+
+        """
+        return self.is_null().if_else(val, self)
```
Comment (jaychia) [inline daft/expressions.py:594]: Just for an additional layer of safety 😛 

```suggestion
        return self.is_null().if_else(lit(val), self)
```

File: daft/expressions.py [hunk 5]
```diff
@@ -570,6 +570,29 @@ def cast(self, to: type) -> Expression:
             )
         raise NotImplementedError(f"Casting to a non-primitive Python object {to} not yet implemented")
 
+    def fillnan(self, val: Any):
```
Comment (jaychia) [inline daft/expressions.py:573]:
```suggestion
    def fillnan(self, val: float | None) -> Expression:
```

File: daft/expressions.py [hunk 6]
```diff
@@ -570,6 +570,29 @@ def cast(self, to: type) -> Expression:
...
+        """Fills any NaN with the value of the val argument.
+        This method can only be applied to float columns!
+
+        Args:
+            val: value which replaces NaN values
+
+        Returns:
+
```
Comment (jaychia) [inline daft/expressions.py:581]:
```suggestion
            Expression: new expression with all NaN values replaced with `val`
```

File: daft/expressions.py [hunk 7]
```diff
@@ -570,6 +570,29 @@ def cast(self, to: type) -> Expression:
...
+        """
+        return self.is_nan().if_else(val, self)
```
Comment (jaychia) [inline daft/expressions.py:583]:
```suggestion
        return self.is_nan().if_else(lit(val), self)
```

**Review by jaychia — COMMENTED** [review]
Awesome stuff!

1. The expression-level `fillnan` and `fillnull` helpers look good!
2. We probably don't want a DataFrame-level `fillnan` and `fillnull`. Users should perform it individually on each column using `df["my_float_col"].fillnan(3.1)` instead. See my comment in `dataframe.py`!

Tests look good, maybe add some corner-case testing:

1. Filling with `None` values
2. For `fillnan`, what happens if a user specifies a non-float `val`? We should make sure that we throw a useful error here. This might already be the case in `if_else`.

Great work, and thanks for the PR! :)

File: tests/dataframe/test_fill.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+import daft.errors
+from daft import DataFrame
+
+
+def test_fill_nan_data_frame(missing_value_data: list[dict[str, Any]]) -> None:
+    df = DataFrame.from_pylist(missing_value_data)
+    df = df.with_column("sepal_width", df["sepal_width"].fillnan(3.1))
+    with pytest.raises(daft.errors.ExpressionTypeError):
+        df = df.with_column("sepal_width_none", df["sepal_width"].fillnan(None))
```
Comment (FelixKleineBoesing) [inline tests/dataframe/test_fill.py:15]: @jaychia None can´t be used with float columns, since it throws an expressiontypeerror. In Pandas None will be converted to NaN if the column is float. 
Do you know how I am able to check the type of a column to assert that this is only applied to float columns? There is the method _resolve_type, but one needs the schema for this. I was thinking to supply the schema if an expression is derived by DataFrame.__getitem__ but that wouldn't solve it for col(column_name) calls.

Comment (jaychia) [reply tests/dataframe/test_fill.py:15]: Ah indeed! This seems to be a corner-case bug that we haven't anticipated I think.

Under the hood `lit(None)` gets a `Null` type, but it should actually take on whatever type it is applied against.

Feel free to add a comment/skip this test for now - in our upcoming Rust release, these will be handled appropriately through type-promotion (essentially, `float + null = float`)

Comment (jaychia) [reply tests/dataframe/test_fill.py:15]: For the second part of your question about asserting that it runs against a float column, this should actually already be handled by your use of `.is_nan()`.

For example, the following snippet will fail:

```
df = DataFrame.from_pydict({"x": ["a", "b", "c"]})

# Throws ExpressionTypeError: Unable to resolve type for operation: is_nan(col(x))
df.with_column("should_fail", df["x"].is_nan())
```

**Comment by samster25** [comment]
@colin-ho do you mind taking a look at this PR and implementing `fill_null` and `fill_nan` for our rust expressions?

**Comment by colin-ho** [comment]
Closed with: https://github.com/Eventual-Inc/Daft/pull/2313
